# Characters Generation

In the first part of the code, we use the necessary functions to generate characters, whose characteristics will be analyzed in the second phase.

Currently, there are two different scripts, each containing two prompts: one for generating positively connotated characters and another for negatively connotated characters.

If necessary for the thesis, a third function will be added that takes as input some characteristics and gives as output the connotation (positive/negative).

In [8]:
import ollama #library that allows us to use Llama locally
import csv
import random
import json
import os

In [2]:
from functions import generate_characters_no_author
generate_characters_no_author(iterations=1) #since I've already generated more than 500 characters, I'll just place this function here as a placeholder

In [3]:
from functions import generate_characters
generate_characters(author=None, iterations=1)

# Characters Analysis

## No Author Analysis

In [11]:
import pandas as pd
import scipy.stats as stats
from scipy.stats import entropy
from scipy.spatial.distance import jensenshannon
from scipy.special import rel_entr
import numpy as np

In [6]:
from functions import process_character_data
process_character_data('characters_no_writer.json', keys_to_extract=None, writer="")

ethnicity data saved to csv\ethnicity_data.csv
Processed ethnicity data saved to csv\processed_ethnicity_data.csv
moral description data saved to csv\moraldescription_data.csv
Processed moral description data saved to csv\processed_moraldescription_data.csv
physical description data saved to csv\physicaldescription_data.csv
Processed physical description data saved to csv\processed_physicaldescription_data.csv
religion data saved to csv\religion_data.csv
Processed religion data saved to csv\processed_religion_data.csv
sex data saved to csv\sex_data.csv
Processed sex data saved to csv\processed_sex_data.csv


### Ethnicity

In [16]:
from functions import process_ethnicity_data
df = process_ethnicity_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_ethnicity_data.csv')
df

ethnicity connotation  count  relative_frequency
42         japanese    negative     99            0.197605
10        caucasian    negative     77            0.153693
80          russian    negative     45            0.089820
78         romanian    negative     39            0.077844
0              arab    negative     30            0.059880
38            irish    negative     30            0.059880
55          kurdish    negative     30            0.059880
12           celtic    negative     21            0.041916
28            gypsy    negative     11            0.021956
73           polish    negative     11            0.021956
26         gujarati    negative      9            0.017964
62          mexican    negative      7            0.013972
75          punjabi    negative      6            0.011976
98            tatar    negative      6            0.011976
43         japanese    positive    181            0.362000
36           indian    positive     92            0.184000
81          russian    positive     39            0.078000
91             sikh    positive     33            0.066000
76          punjabi    positive     32            0.064000
56          kurdish    positive     16            0.032000
25            greek    positive     15            0.030000
27         gujarati    positive     11            0.022000
93           somali    positive      9            0.018000
46         kashmiri    positive      6            0.012000
22  french-canadian    positive      5            0.010000
71          persian    positive      5            0.010000
83            rusyn    positive      5            0.010000
86     scandinavian    positive      5            0.010000

As we can see, Japanese appears most frequently in both negative (99 instances, 19.76%) and positive (181 instances, 36.2%) categories, suggesting a strong but polarized representation. Other ethnicities such as Caucasian, Russian, and Punjabi also appear in both connotations but with varying frequencies. It's pretty strange how many japanese characters are generated, compared to other groups.

In [17]:
from functions import chi_square_test
result = chi_square_test(df, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(453.0931),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 22}

The chi-square test results indicate a statistically significant relationship between ethnicity and connotation (positive/negative):

Chi-Square Statistic = 453.0931 \
Degrees of Freedom = 22 \
p-value = 0.0 (effectively zero) 

Since the p-value is far below the standard significance threshold, we reject the null hypothesis, meaning that the distribution of positive and negative connotations across ethnicities is not random. This strongly suggests the presence of bias in how different ethnicities are associated with positive or negative adjectives in the dataset. Further investigation will help identify which ethnicities deviate most from expected distributions.

In [18]:
from functions import compute_standardized_residuals
result = compute_standardized_residuals(df, index_col='ethnicity')
result

connotation  negative  positive
ethnicity                      
arab         4.097051 -3.945341
caucasian    6.563812 -6.320759
celtic       3.427839 -3.300909
greek       -2.686474  2.586996
gypsy        2.480888 -2.389023
indian      -6.653205  6.406843
irish        4.097051 -3.945341
japanese    -3.077481  2.963525
polish       2.480888 -2.389023
punjabi     -2.872703  2.766329
romanian     4.671357 -4.498381
sikh        -3.984685  3.837135
somali      -2.080934  2.003879

These standardized residuals from the chi-square test indicate which ethnicities deviate significantly from expected distributions of positive and negative connotations. Key observations:

- Overrepresentation of negativity (positive residuals in the "negative" column, negative residuals in the "positive" column): 
    - Caucasian (6.56, -6.32), Romanian (4.67, -4.50), Irish (4.10, -3.95), Arab (4.10, -3.95), Gypsy (2.48, -2.39), Polish (2.48, -2.39), Celtic (3.43, -3.30) \
      These groups are disproportionately associated with negative connotations.

- Overrepresentation of positivity (negative residuals in the "negative" column, positive residuals in the "positive" column):
    - Indian (-6.65, 6.41), Sikh (-3.98, 3.84), Greek (-2.69, 2.59), Punjabi (-2.87, 2.77), Somali (-2.08, 2.00) \
      These groups are disproportionately associated with positive connotations.

- Mixed cases (Japanese (-3.08, 2.96)) show more balanced but still significant deviations.

Interpretation: \
These deviations suggest bias in sentiment distribution, with some ethnicities systematically skewed toward positive or negative characterizations. \
The strongest disparities appear for Caucasian (negative) and Indian (positive), indicating areas where bias might be particularly pronounced. 

In [19]:
from functions import compute_divergences
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='ethnicity')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.709721470703967)},
 'KL': {'KL(negative || positive)': np.float64(11.236501846761943),
  'KL(positive || negative)': np.float64(8.19796023576817)}}

These statistical measures provide insight into the **divergence** between the distributions of positive and negative connotations across ethnicities:

### **Jensen-Shannon Divergence (JSD)**
- **JSD(negative || positive) = 0.7097**  
  - JSD is a symmetric measure of distributional similarity, ranging from 0 (identical distributions) to 1 (completely different distributions).
  - A value of **0.7097** suggests a **high degree of divergence** between the distributions of negative and positive connotations. This indicates that the way positive and negative adjectives are distributed across ethnicities is **substantially different**.

### **Kullback-Leibler Divergence (KL)**
- **KL(negative || positive) = 11.2365**  
- **KL(positive || negative) = 8.1980**  
  - KL divergence is an asymmetric measure of how much one distribution diverges from another.
  - The fact that **KL(negative || positive) > KL(positive || negative)** suggests that the negative connotation distribution is more different from the positive distribution than vice versa. This could indicate that negative connotations are more concentrated among certain ethnicities, whereas positive connotations might be more evenly distributed.

### **Interpretation**
- The high **JSD** and **KL** values confirm that **ethnicity plays a strong role in shaping the sentiment distribution**, reinforcing the bias seen in the chi-square test.
- The asymmetry in **KL divergence** suggests that the **negative connotations may be more skewed** toward certain ethnicities than the positive ones.

### Sex

In [21]:
from functions import process_sex_data
df = process_sex_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_sex_data.csv')
df

sex connotation  count  relative_frequency
2    male    negative    497            0.992016
0  female    negative      4            0.007984
1  female    positive    297            0.594000
3    male    positive    203            0.406000

In [22]:
result = chi_square_test(df, index_col='sex')
result

{'Chi-Square Statistic': np.float64(405.9099),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 1}

In [23]:
result = compute_standardized_residuals(df, index_col='sex')
result

connotation   negative   positive
sex                              
female      -11.948077  11.960020
male          7.834878  -7.842709

In [24]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='sex')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.6014721467174361)},
 'KL': {'KL(negative || positive)': np.float64(0.8518465487854805),
  'KL(positive || negative)': np.float64(2.197090119501996)}}

### Religion

In [25]:
from functions import process_religion_data
df = process_religion_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_religion_data.csv')
df

religion connotation  count  relative_frequency
3            atheism    negative    117            0.253247
38         shintoism    negative     81            0.175325
21          islamism    negative     57            0.123377
28       orthodoxism    negative     45            0.097403
10      christianity    negative     36            0.077922
8        catholicism    negative     29            0.062771
5            atheist    negative     21            0.045455
30          paganism    negative     19            0.041126
14          hinduism    negative     16            0.034632
43        sunniislam    negative      9            0.019481
6           buddhism    negative      7            0.015152
32  romancatholicism    negative      5            0.010823
34           secular    negative      5            0.010823
39         shintoism    positive    144            0.288000
15          hinduism    positive    139            0.278000
29       orthodoxism    positive     62            0.124000
7           buddhism    positive     42            0.084000
41           sikhism    positive     37            0.074000
22          islamism    positive     28            0.056000
9        catholicism    positive     10            0.020000
11      christianity    positive     10            0.020000
50    zoroastrianism    positive      5            0.010000

In [26]:
result = chi_square_test(df, index_col='religion')
result

{'Chi-Square Statistic': np.float64(394.2351),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 14}

In [27]:
result = compute_standardized_residuals(df, index_col='religion')
result

connotation   negative  positive
religion                        
atheism       8.028262 -7.771702
atheist       3.401248 -3.292554
buddhism     -3.430985  3.321341
catholicism   2.332883 -2.258331
christianity  2.914093 -2.820967
hinduism     -6.811596  6.593917
islamism      2.476394 -2.397255
paganism      3.235232 -3.131843
shintoism    -2.669166  2.583867
sikhism      -4.230762  4.095559
sunniislam    2.226639 -2.155482

In [28]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='religion')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.6219273103270238)},
 'KL': {'KL(negative || positive)': np.float64(8.318515404850475),
  'KL(positive || negative)': np.float64(2.6343018650557717)}}

### Physical

In [29]:
from functions import process_phy_data
df = process_phy_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_physicaldescription_data.csv')
df

physicaldescription connotation  count  relative_frequency
405             scrawny    negative    241            0.096671
332                pale    negative    232            0.093061
504                tall    negative    213            0.085439
266                lean    negative    169            0.067790
562           weathered    negative    105            0.042118
..                  ...         ...    ...                 ...
520                thin    positive     14            0.005632
144                eyes    positive     13            0.005229
235      highcheekbones    positive     13            0.005229
345         piercedeyes    positive     13            0.005229
448               small    positive     13            0.005229

[77 rows x 4 columns]

In [30]:
result = chi_square_test(df, index_col='physicaldescription')
result

{'Chi-Square Statistic': np.float64(2652.3951),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 68}

In [31]:
result = compute_standardized_residuals(df, index_col='physicaldescription')
result

connotation          negative  positive
physicaldescription                    
balding              4.675393 -4.685399
bearded              3.152149 -3.158896
bony                 3.229992 -3.236905
bright              -2.918591  2.924837
bright-eyed         -2.741539  2.747406
...                       ...       ...
weathered            7.222482 -7.237939
wizened              2.541344 -2.546783
worn                 3.306002 -3.313078
young               -5.390918  5.402456
youthful            -4.363556  4.372895

[64 rows x 2 columns]

In [32]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='physicaldescription')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.8358405401706462)},
 'KL': {'KL(negative || positive)': np.float64(11.924524392232907),
  'KL(positive || negative)': np.float64(14.029306987763642)}}

### Moral

In [34]:
from functions import process_mor_data
df = process_mor_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_moraldescription_data.csv')
df

moraldescription connotation  count  relative_frequency
117     manipulative    negative    344            0.137600
31             cruel    negative    335            0.134000
151         ruthless    negative    171            0.068400
155          selfish    negative    155            0.062000
32           cunning    negative    146            0.058400
135     power-hungry    negative    139            0.055600
1          ambitious    negative    101            0.040400
36         deceitful    negative     95            0.038000
141         reckless    negative     72            0.028800
181       vindictive    negative     56            0.022400
37         deceptive    negative     50            0.020000
178          vicious    negative     50            0.020000
73         heartless    negative     48            0.019200
3             amoral    negative     40            0.016000
62      exploitative    negative     40            0.016000
71            greedy    negative     39            0.015600
16           callous    negative     36            0.014400
21              cold    negative     36            0.014400
173    untrustworthy    negative     36            0.014400
175         vengeful    negative     34            0.013600
182          violent    negative     29            0.011600
51         dishonest    negative     26            0.010400
152         sadistic    negative     25            0.010000
23      cold-hearted    negative     21            0.008400
113            lying    negative     21            0.008400
153         scheming    negative     20            0.008000
10        avaricious    negative     19            0.007600
43       destructive    negative     16            0.006400
118        merciless    negative     16            0.006400
120     narcissistic    negative     15            0.006000
7           arrogant    negative     13            0.005200
54          disloyal    negative     13            0.005200
25     compassionate    positive    469            0.187901
144        resilient    positive    245            0.098157
93       intelligent    positive    191            0.076522
108            loyal    positive    170            0.068109
44        determined    positive    134            0.053686
60        empathetic    positive    134            0.053686
27        courageous    positive    130            0.052083
67          generous    positive     97            0.038862
33           curious    positive     79            0.031651
131          patient    positive     73            0.029247
74            honest    positive     70            0.028045
147      resourceful    positive     69            0.027644
8          authentic    positive     61            0.024439
97              kind    positive     52            0.020833
148       respectful    positive     50            0.020032
30          creative    positive     43            0.017228
94     introspective    positive     43            0.017228
81       independent    positive     42            0.016827
2          ambitious    positive     33            0.013221
78            humble    positive     33            0.013221
156         selfless    positive     30            0.012019
75         honorable    positive     29            0.011619
13             brave    positive     26            0.010417
189             wise    positive     23            0.009215
85       inquisitive    positive     21            0.008413
38         dedicated    positive     17            0.006811

In [35]:
result = chi_square_test(df, index_col='moraldescription')
result

{'Chi-Square Statistic': np.float64(4521.4541),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 56}

In [36]:
result = compute_standardized_residuals(df, index_col='moraldescription')
result

connotation        negative   positive
moraldescription                      
ambitious          4.394470  -4.293866
amoral             4.629603  -4.523616
arrogant           2.639280  -2.578858
authentic         -5.458367   5.333407
avaricious         3.190733  -3.117687
brave             -3.563563   3.481982
callous            4.392027  -4.291479
cold               4.392027  -4.291479
cold-hearted       3.354466  -3.277671
compassionate    -15.135062  14.788573
courageous        -7.968370   7.785948
creative          -4.582812   4.477897
cruel             13.397881 -13.091161
cunning            8.844843  -8.642357
curious           -6.211712   6.069507
deceitful          7.134697  -6.971361
deceptive          5.176053  -5.057557
dedicated         -2.881524   2.815557
destructive        2.928018  -2.860986
determined        -8.090031   7.904825
dishonest          3.732505  -3.647056
disloyal           2.639280  -2.578858
empathetic        -8.090031   7.904825
exploitative       4.629603  -4.523616
generous          -6.883093   6.725517
greedy             4.571366  -4.466713
heartless          5.071475  -4.955373
honest            -5.847185   5.713324
honorable         -3.763542   3.677383
humble            -4.014715   3.922806
independent       -4.529210   4.425522
inquisitive       -3.202635   3.129316
intelligent       -9.658607   9.437490
introspective     -4.582812   4.477897
kind              -5.039639   4.924266
loyal             -9.112179   8.903572
lying              3.354466  -3.277671
manipulative      13.576660 -13.265847
merciless          2.928018  -2.860986
narcissistic       2.835041  -2.770138
patient           -5.971167   5.834468
power-hungry       8.630205  -8.432632
reckless           6.211264  -6.069068
resilient        -10.939081  10.688651
resourceful       -5.805269   5.672368
respectful        -4.941773   4.828640
ruthless           9.572200  -9.353062
sadistic           3.660022  -3.576233
scheming           3.273623  -3.198680
selfish            9.113382  -8.904747
selfless          -3.827881   3.740249
untrustworthy      4.392027  -4.291479
vengeful           4.268283  -4.170568
vicious            5.176053  -5.057557
vindictive         5.477820  -5.352415
violent            3.941964  -3.851720
wise              -3.351674   3.274943

In [37]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='moraldescription')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.988317402502221)},
 'KL': {'KL(negative || positive)': np.float64(19.200290844548718),
  'KL(positive || negative)': np.float64(19.87708117457782)}}

## Author Analysis - Hemingway Tests

In [38]:
process_character_data('characters_output.json', keys_to_extract=None, writer="Ernest Hemingway")

ethnicity data saved to csv\ethnicity_data_hemingway.csv
Processed ethnicity data saved to csv\processed_ethnicity_data_hemingway.csv
moral description data saved to csv\moraldescription_data_hemingway.csv
Processed moral description data saved to csv\processed_moraldescription_data_hemingway.csv
physical description data saved to csv\physicaldescription_data_hemingway.csv
Processed physical description data saved to csv\processed_physicaldescription_data_hemingway.csv
religion data saved to csv\religion_data_hemingway.csv
Processed religion data saved to csv\processed_religion_data_hemingway.csv
sex data saved to csv\sex_data_hemingway.csv
Processed sex data saved to csv\processed_sex_data_hemingway.csv


### Ethnicity

In [39]:
df = process_ethnicity_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_ethnicity_data_hemingway.csv')
df

ethnicity connotation  count  relative_frequency
50           mexican    negative    144            0.360902
31          japanese    negative     68            0.170426
63           russian    negative     50            0.125313
52  mexican-american    negative     11            0.027569
71           spanish    negative     10            0.025063
28             irish    negative      9            0.022556
22          hispanic    negative      8            0.020050
25            indian    negative      8            0.020050
41    latinoamerican    negative      8            0.020050
15             cuban    negative      7            0.017544
6             basque    negative      5            0.012531
39           kurdish    negative      5            0.012531
48           mestizo    negative      5            0.012531
30           italian    negative      4            0.010025
46             mayan    negative      4            0.010025
32          japanese    positive     62            0.305419
51           mexican    positive     58            0.285714
47             mayan    positive      8            0.039409
7             basque    positive      6            0.029557
23          hispanic    positive      5            0.024631
26            indian    positive      5            0.024631
16             cuban    positive      4            0.019704
29             irish    positive      4            0.019704
37            korean    positive      4            0.019704
42    latinoamerican    positive      4            0.019704
49           mestizo    positive      4            0.019704
53  mexican-american    positive      4            0.019704
64           russian    positive      4            0.019704
72           spanish    positive      4            0.019704
40           kurdish    positive      3            0.014778
45              maya    positive      3            0.014778
56    nativeamerican    positive      3            0.014778

In [40]:
result = chi_square_test(df, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(60.1201),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 17}

In [41]:
result = compute_standardized_residuals(df, index_col='ethnicity')
result

connotation  negative  positive
ethnicity                      
russian      2.497306 -3.415261

In [42]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='ethnicity')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.31554838434512783)},
 'KL': {'KL(negative || positive)': np.float64(0.48617537425936325),
  'KL(positive || negative)': np.float64(1.1607051948223843)}}

### Sex

In [43]:
df = process_sex_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_sex_data_hemingway.csv')
df

sex connotation  count  relative_frequency
2    male    negative    398            0.997494
0  female    negative      1            0.002506
3    male    positive    200            0.985222
1  female    positive      3            0.014778

In [44]:
result = chi_square_test(df, index_col='sex')
result

{'Chi-Square Statistic': np.float64(1.4922),
 'p-value': np.float64(0.2219),
 'Degrees of Freedom': 1}

In [45]:
result = compute_standardized_residuals(df, index_col='sex')
result

Empty DataFrame
Columns: [negative, positive]
Index: []

In [46]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='sex')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.05923379207262645)},
 'KL': {'KL(negative || positive)': np.float64(0.007901138413626807),
  'KL(positive || negative)': np.float64(0.01402592611765316)}}

### Religion

In [47]:
df = process_religion_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_religion_data_hemingway.csv')
df

religion connotation  count  relative_frequency
7              catholicism    negative    151            0.467492
2                  atheism    negative     67            0.207430
26               shintoism    negative     36            0.111455
9             christianity    negative     15            0.046440
19             orthodoxism    negative     15            0.046440
4                  atheist    negative     11            0.034056
16                islamism    negative      8            0.024768
22        romancatholicism    negative      5            0.015480
8              catholicism    positive    104            0.541667
27               shintoism    positive     56            0.291667
3                  atheism    positive      8            0.041667
20             orthodoxism    positive      6            0.031250
13              indigenous    positive      3            0.015625
23        romancatholicism    positive      3            0.015625
6                 buddhism    positive      2            0.010417
10            christianity    positive      2            0.010417
14  indigenousspirituality    positive      2            0.010417
17                islamism    positive      2            0.010417

In [48]:
result = chi_square_test(df, index_col='religion')
result

{'Chi-Square Statistic': np.float64(70.4113),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 10}

In [49]:
result = compute_standardized_residuals(df, index_col='religion')
result

connotation  negative  positive
religion                       
atheism      2.993287 -3.831288
shintoism   -2.795446  3.578060

In [50]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='religion')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.3563017388548288)},
 'KL': {'KL(negative || positive)': np.float64(1.0078512932526835),
  'KL(positive || negative)': np.float64(0.9312442201903928)}}

### Physical

In [51]:
df = process_phy_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_physicaldescription_data_hemingway.csv')
df

physicaldescription connotation  count  relative_frequency
405           weathered    negative    209            0.104762
365                tall    negative    199            0.099749
303             scrawny    negative    139            0.069674
205                lean    negative    138            0.069173
282             scarred    negative     88            0.044110
294               scars    negative     71            0.035589
62          crookednose    negative     63            0.031579
364          sunkeneyes    negative     63            0.031579
327              strong    negative     56            0.028070
377            tattooed    negative     33            0.016541
410       weatheredface    negative     29            0.014536
6               balding    negative     27            0.013534
414       weatheredskin    negative     23            0.011529
228            muscular    negative     22            0.011028
239                pale    negative     22            0.011028
287         scarredface    negative     22            0.011028
338       strongjawline    negative     22            0.011028
387                thin    negative     22            0.011028
25                broad    negative     21            0.010526
147               gaunt    negative     18            0.009023
381             tattoos    negative     18            0.009023
421                worn    negative     18            0.009023
150             gnarled    negative     17            0.008521
71             darkeyes    negative     15            0.007519
354       sun-weathered    negative     15            0.007519
270              sallow    negative     13            0.006516
299         scarsonface    negative     12            0.006015
58              crooked    negative     11            0.005514
191         hollow-eyed    negative     11            0.005514
211            leathery    negative     11            0.005514
215        leatheryskin    negative     11            0.005514
30       broadshoulders    negative     10            0.005013
366                tall    positive    140            0.137931
406           weathered    positive    119            0.117241
206                lean    positive     99            0.097537
328              strong    positive     91            0.089655
339       strongjawline    positive     53            0.052217
411       weatheredface    positive     36            0.035468
415       weatheredskin    positive     36            0.035468
26                broad    positive     25            0.024631
229            muscular    positive     25            0.024631
31       broadshoulders    positive     23            0.022660
295               scars    positive     22            0.021675
283             scarred    positive     20            0.019704
72             darkeyes    positive     19            0.018719
74             darkhair    positive     15            0.014778
174             haggard    positive     11            0.010837
288         scarredface    positive     10            0.009852
28     broad-shouldered    positive      9            0.008867
302        scarsonhands    positive      9            0.008867
335         stronghands    positive      9            0.008867
45            browneyes    positive      8            0.007882
69                 dark    positive      8            0.007882
300         scarsonface    positive      7            0.006897
165            greyhair    positive      6            0.005911
216        leatheryskin    positive      6            0.005911
330        strong-jawed    positive      6            0.005911
425           wornboots    positive      6            0.005911

In [52]:
result = chi_square_test(df, index_col='physicaldescription')
result

{'Chi-Square Statistic': np.float64(607.6384),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 41}

In [53]:
result = chi_square_test(df, index_col='physicaldescription')
result

{'Chi-Square Statistic': np.float64(607.6384),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 41}

In [54]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='physicaldescription')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.530114657843018)},
 'KL': {'KL(negative || positive)': np.float64(6.659136311885535),
  'KL(positive || negative)': np.float64(2.3907304977845696)}}

### Moral

In [55]:
df = process_mor_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\processed_moraldescription_data_hemingway.csv')
df

moraldescription connotation  count  relative_frequency
236          selfish    negative    127            0.063659
37             cruel    negative    123            0.061654
171     manipulative    negative     94            0.047118
136      independent    negative     87            0.043609
112          haunted    negative     76            0.038095
..               ...         ...    ...                 ...
234     self-reliant    positive      7            0.006897
237          selfish    positive      7            0.006897
244          sincere    positive      7            0.006897
135        impulsive    positive      6            0.005911
194          private    positive      6            0.005911

[76 rows x 4 columns]

In [56]:
result = chi_square_test(df, index_col='moraldescription')
result

{'Chi-Square Statistic': np.float64(1241.3822),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 59}

In [57]:
result = compute_standardized_residuals(df, index_col='moraldescription')
result

connotation       negative  positive
moraldescription                    
amoral            2.306955 -3.101302
brave            -4.011780  5.393144
cold              2.129223 -2.862372
courageous       -2.892936  3.889052
cowardly          2.263830 -3.043328
cruel             4.923906 -6.619340
cynical           2.736837 -3.679205
deceitful         2.390871 -3.214113
disciplined      -3.497389  4.701634
drysenseofhumor  -2.122834  2.853784
honest           -3.641355  4.895172
honorable        -2.122834  2.853784
humble           -2.537272  3.410924
independent      -4.280452  5.754328
lonely            2.736837 -3.679205
loyal            -5.392375  7.249116
manipulative      4.304484 -5.786635
patient          -2.892936  3.889052
practical        -2.407068  3.235887
reckless          2.807936 -3.774784
resilient        -6.251775  8.404430
resourceful      -4.521383  6.078217
respectful       -4.169164  5.604720
ruthless          3.870472 -5.203181
self-disciplined -2.122834  2.853784
self-reliant     -2.122834  2.853784
selfish           4.385701 -5.895817
selfless         -2.661114  3.577407
sincere          -2.122834  2.853784
vengeful          3.107815 -4.177921
vicious           2.263830 -3.043328
vindictive        3.262527 -4.385903
violent           2.842818 -3.821678
witty            -2.407068  3.235887

In [58]:
connotations = df["connotation"].unique()
result = compute_divergences(df, connotations, index_col='moraldescription')
result

{'JSD': {'JSD(negative || positive)': np.float64(0.7128166249416509)},
 'KL': {'KL(negative || positive)': np.float64(10.744262781031955),
  'KL(positive || negative)': np.float64(4.553780662546629)}}

## Author vs No Author

In [59]:
process_character_data('characters_output.json', keys_to_extract=None, writer="Ernest Hemingway")
process_character_data('characters_no_writer.json', keys_to_extract=None, writer="")

ethnicity data saved to csv\ethnicity_data_hemingway.csv
Processed ethnicity data saved to csv\processed_ethnicity_data_hemingway.csv
moral description data saved to csv\moraldescription_data_hemingway.csv
Processed moral description data saved to csv\processed_moraldescription_data_hemingway.csv
physical description data saved to csv\physicaldescription_data_hemingway.csv
Processed physical description data saved to csv\processed_physicaldescription_data_hemingway.csv
religion data saved to csv\religion_data_hemingway.csv
Processed religion data saved to csv\processed_religion_data_hemingway.csv
sex data saved to csv\sex_data_hemingway.csv
Processed sex data saved to csv\processed_sex_data_hemingway.csv
ethnicity data saved to csv\ethnicity_data.csv
Processed ethnicity data saved to csv\processed_ethnicity_data.csv
moral description data saved to csv\moraldescription_data.csv
Processed moral description data saved to csv\processed_moraldescription_data.csv
physical description data sa

In [61]:
author = process_ethnicity_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\Processed_ethnicity_data_hemingway.csv')
no_author = process_ethnicity_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\Processed_ethnicity_data.csv')

In [62]:
no_author['author'] = 'No Author'
author['author'] = 'Author'

### Ethnicity Distribution

In [78]:
df_list = [author, no_author]
df_concat = pd.concat(df_list)
df_grouped = df_concat.groupby(["ethnicity", "author"])["count"].sum().reset_index()
df_grouped

ethnicity     author  count
0               arab  No Author     30
1             basque     Author     11
2          caucasian  No Author     77
3             celtic  No Author     21
4              cuban     Author     11
5    french-canadian  No Author      5
6              greek  No Author     15
7           gujarati  No Author     20
8              gypsy  No Author     11
9           hispanic     Author     13
10            indian     Author     13
11            indian  No Author     92
12             irish     Author     13
13             irish  No Author     30
14           italian     Author      4
15          japanese     Author    130
16          japanese  No Author    280
17          kashmiri  No Author      6
18            korean     Author      4
19           kurdish     Author      8
20           kurdish  No Author     46
21    latinoamerican     Author     12
22              maya     Author      3
23             mayan     Author     12
24           mestizo     Author      9
25           mexican     Author    202
26           mexican  No Author      7
27  mexican-american     Author     15
28    nativeamerican     Author      3
29           persian  No Author      5
30            polish  No Author     11
31           punjabi  No Author     38
32          romanian  No Author     39
33           russian     Author     54
34           russian  No Author     84
35             rusyn  No Author      5
36      scandinavian  No Author      5
37              sikh  No Author     33
38            somali  No Author      9
39           spanish     Author     14
40             tatar  No Author      6

In [79]:
from functions import chi_square_test_authors
result = chi_square_test_authors(df_grouped, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(743.5843),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 34}

In [80]:
from functions import compute_standardized_residuals_authors
result = compute_standardized_residuals_authors(df_grouped, index_col='ethnicity')
result

author               Author  No Author
ethnicity                             
arab              -3.366009   2.622156
basque             3.358648  -2.616421
caucasian         -5.392622   4.200908
celtic            -2.816205   2.193853
cuban              3.358648  -2.616421
gujarati          -2.748335   2.140981
hispanic           3.651235  -2.844350
indian            -4.232824   3.297414
kurdish           -2.744485   2.137982
latinoamerican     3.507993  -2.732763
mayan              3.507993  -2.732763
mestizo            3.038011  -2.366642
mexican           13.852108 -10.790936
mexican-american   3.922056  -3.055322
punjabi           -3.788318   2.951139
romanian          -3.837840   2.989717
sikh              -3.530300   2.750140
spanish            3.789066  -2.951722

In [83]:
from functions import compute_divergences_authors
authors = df_grouped["author"].unique()
df_grouped["relative_frequency"] = df_grouped.groupby("author")["count"].transform(lambda x: x / x.sum())
result = compute_divergences_authors(df_grouped, authors, index_col='ethnicity')
result

{'JSD': {'JSD(No Author || Author)': np.float64(0.7030834032147633)},
 'KL': {'KL(No Author || Author)': np.float64(7.8045888919229),
  'KL(Author || No Author)': np.float64(5.339351395791128)}}

### Positive Connotation - Ethnicity

In [63]:
no_author_pos = no_author[no_author['connotation'] == 'positive']
author_pos = author[author['connotation'] == 'positive']

In [67]:
df_list = [author_pos, no_author_pos]
df_concat = pd.concat(df_list)
df_concat

ethnicity connotation  count  relative_frequency     author
32          japanese    positive     62            0.305419     Author
51           mexican    positive     58            0.285714     Author
47             mayan    positive      8            0.039409     Author
7             basque    positive      6            0.029557     Author
23          hispanic    positive      5            0.024631     Author
26            indian    positive      5            0.024631     Author
16             cuban    positive      4            0.019704     Author
29             irish    positive      4            0.019704     Author
37            korean    positive      4            0.019704     Author
42    latinoamerican    positive      4            0.019704     Author
49           mestizo    positive      4            0.019704     Author
53  mexican-american    positive      4            0.019704     Author
64           russian    positive      4            0.019704     Author
72           spanish    positive      4            0.019704     Author
40           kurdish    positive      3            0.014778     Author
45              maya    positive      3            0.014778     Author
56    nativeamerican    positive      3            0.014778     Author
43          japanese    positive    181            0.362000  No Author
36            indian    positive     92            0.184000  No Author
81           russian    positive     39            0.078000  No Author
91              sikh    positive     33            0.066000  No Author
76           punjabi    positive     32            0.064000  No Author
56           kurdish    positive     16            0.032000  No Author
25             greek    positive     15            0.030000  No Author
27          gujarati    positive     11            0.022000  No Author
93            somali    positive      9            0.018000  No Author
46          kashmiri    positive      6            0.012000  No Author
22   french-canadian    positive      5            0.010000  No Author
71           persian    positive      5            0.010000  No Author
83             rusyn    positive      5            0.010000  No Author
86      scandinavian    positive      5            0.010000  No Author

In [69]:
result = chi_square_test_authors(df_concat, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(361.515),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 26}

In [70]:
result = compute_standardized_residuals_authors(df_concat, index_col='ethnicity')
result

author        Author  No Author
ethnicity                      
basque      3.234411  -2.064682
indian     -4.355819   2.780531
mayan       3.734776  -2.384089
mexican    10.056192  -6.419356

In [71]:
authors = df_concat["author"].unique()
result = compute_divergences_authors(df_concat, authors, index_col='ethnicity')
result

{'JSD': {'JSD(Author || No Author)': np.float64(0.7162417055910915)},
 'KL': {'KL(Author || No Author)': np.float64(12.23822955692644),
  'KL(No Author || Author)': np.float64(6.099929452122382)}}

### Negative Connotation  - Ethnicity

In [72]:
no_author_neg = no_author[no_author['connotation'] == 'negative']
author_neg = author[author['connotation'] == 'negative']

In [73]:
df_list = [author_neg, no_author_neg]
df_concat = pd.concat(df_list)
df_concat

ethnicity connotation  count  relative_frequency     author
50           mexican    negative    144            0.360902     Author
31          japanese    negative     68            0.170426     Author
63           russian    negative     50            0.125313     Author
52  mexican-american    negative     11            0.027569     Author
71           spanish    negative     10            0.025063     Author
28             irish    negative      9            0.022556     Author
22          hispanic    negative      8            0.020050     Author
25            indian    negative      8            0.020050     Author
41    latinoamerican    negative      8            0.020050     Author
15             cuban    negative      7            0.017544     Author
6             basque    negative      5            0.012531     Author
39           kurdish    negative      5            0.012531     Author
48           mestizo    negative      5            0.012531     Author
30           italian    negative      4            0.010025     Author
46             mayan    negative      4            0.010025     Author
42          japanese    negative     99            0.197605  No Author
10         caucasian    negative     77            0.153693  No Author
80           russian    negative     45            0.089820  No Author
78          romanian    negative     39            0.077844  No Author
0               arab    negative     30            0.059880  No Author
38             irish    negative     30            0.059880  No Author
55           kurdish    negative     30            0.059880  No Author
12            celtic    negative     21            0.041916  No Author
28             gypsy    negative     11            0.021956  No Author
73            polish    negative     11            0.021956  No Author
26          gujarati    negative      9            0.017964  No Author
62           mexican    negative      7            0.013972  No Author
75           punjabi    negative      6            0.011976  No Author
98             tatar    negative      6            0.011976  No Author

In [74]:
result = chi_square_test_authors(df_concat, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(436.3186),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 23}

In [75]:
result = compute_standardized_residuals_authors(df_concat, index_col='ethnicity')
result

author              Author  No Author
ethnicity                            
arab             -3.678756   3.335017
caucasian        -5.893669   5.342970
celtic           -3.077868   2.790275
gypsy            -2.227597   2.019453
hispanic          2.311486  -2.095503
indian            2.311486  -2.095503
kurdish          -2.715178   2.461474
latinoamerican    2.311486  -2.095503
mexican           9.194195  -8.335098
mexican-american  2.710458  -2.457195
polish           -2.227597   2.019453
romanian         -4.194427   3.802504
spanish           2.584320  -2.342844

In [76]:
authors = df_concat["author"].unique()
result = compute_divergences_authors(df_concat, authors, index_col='ethnicity')
result

{'JSD': {'JSD(Author || No Author)': np.float64(0.7362750744707991)},
 'KL': {'KL(Author || No Author)': np.float64(5.17976215858654),
  'KL(No Author || Author)': np.float64(10.364799326053703)}}

## Author vs Author

In [3]:
from functions import process_character_data
process_character_data('characters_output.json', keys_to_extract=None, writer="Edgar Allan Poe")
process_character_data('characters_output.json', keys_to_extract=None, writer="Jane Austen")

ethnicity data saved to csv\ethnicity_data_poe.csv
Processed ethnicity data saved to csv\processed_ethnicity_data_poe.csv
moral description data saved to csv\moraldescription_data_poe.csv
Processed moral description data saved to csv\processed_moraldescription_data_poe.csv
physical description data saved to csv\physicaldescription_data_poe.csv
Processed physical description data saved to csv\processed_physicaldescription_data_poe.csv
religion data saved to csv\religion_data_poe.csv
Processed religion data saved to csv\processed_religion_data_poe.csv
sex data saved to csv\sex_data_poe.csv
Processed sex data saved to csv\processed_sex_data_poe.csv
ethnicity data saved to csv\ethnicity_data_austen.csv
Processed ethnicity data saved to csv\processed_ethnicity_data_austen.csv
moral description data saved to csv\moraldescription_data_austen.csv
Processed moral description data saved to csv\processed_moraldescription_data_austen.csv
physical description data saved to csv\physicaldescription_d

In [5]:
from functions import process_ethnicity_data
poe = process_ethnicity_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\Processed_ethnicity_data_poe.csv')
austen = process_ethnicity_data(r'C:\Users\edoar\Documents\Tesi\thesis\csv\Processed_ethnicity_data_austen.csv')

In [6]:
poe['author'] = 'Poe'
austen['author'] = 'Austen'

### Ethnicity Distribution

In [12]:
df_list = [poe, austen]
df_concat = pd.concat(df_list)
df_grouped = df_concat.groupby(["ethnicity", "author"])["count"].sum().reset_index()
df_grouped

ethnicity  author  count
0            afghan  Austen      8
1           african  Austen     77
2           african     Poe      3
3   africanamerican  Austen      8
4   africanamerican     Poe     32
5              arab     Poe      6
6             black  Austen      6
7     britishindian  Austen      4
8             cajun     Poe     21
9         caucasian  Austen     31
10        caucasian     Poe     46
11           celtic     Poe     10
12             cree  Austen      9
13            cuban     Poe      3
14          english  Austen     71
15    englishindian  Austen     41
16   englishmestiza  Austen      3
17  french-canadian     Poe     20
18            greek     Poe      7
19            gypsy     Poe      3
20         hispanic     Poe      6
21           indian  Austen    153
22           indian     Poe     21
23            irish  Austen     27
24            irish     Poe    104
25          italian     Poe      3
26         japanese     Poe      8
27           korean  Austen      4
28          kurdish  Austen     32
29          kurdish     Poe     18
30   nativeamerican  Austen      3
31   nativeamerican     Poe      3
32          persian     Poe     15
33          punjabi     Poe      5
34         romanian  Austen      6
35          russian     Poe      7
36         scottish  Austen     13
37   scottishgaelic  Austen      6
38             sikh     Poe      6
39          spanish     Poe      3

In [13]:
from functions import chi_square_test_authors
result = chi_square_test_authors(df_grouped, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(518.4741),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 32}

In [14]:
from functions import compute_standardized_residuals_authors
result = compute_standardized_residuals_authors(df_grouped, index_col='ethnicity')
result

author             Austen       Poe
ethnicity                          
african          4.349794 -5.209386
africanamerican -3.206807  3.840526
cajun           -3.517562  4.212690
caucasian       -2.133219  2.554778
celtic          -2.427348  2.907033
english          4.509471 -5.400617
englishindian    3.426797 -4.103989
french-canadian -3.432789  4.111165
greek           -2.030865  2.432198
indian           4.985434 -5.970639
irish           -5.712288  6.841131
japanese        -2.171086  2.600129
persian         -2.972882  3.560373
russian         -2.030865  2.432198

In [15]:
from functions import compute_divergences_authors
authors = df_grouped["author"].unique()
df_grouped["relative_frequency"] = df_grouped.groupby("author")["count"].transform(lambda x: x / x.sum())
result = compute_divergences_authors(df_grouped, authors, index_col='ethnicity')
result

{'JSD': {'JSD(Austen || Poe)': np.float64(0.7521262230331562)},
 'KL': {'KL(Austen || Poe)': np.float64(7.64613464393532),
  'KL(Poe || Austen)': np.float64(7.483859300977419)}}

### Positive - Ethnicity

In [16]:
poe_pos = poe[poe['connotation'] == 'positive']
austen_pos = austen[austen['connotation'] == 'positive']

In [17]:
df_list = [poe_pos, austen_pos]
df_concat = pd.concat(df_list)
df_concat

ethnicity connotation  count  relative_frequency  author
33           indian    positive     21            0.104478     Poe
25  french-canadian    positive     16            0.079602     Poe
35            irish    positive     16            0.079602     Poe
13        caucasian    positive     15            0.074627     Poe
54          persian    positive     15            0.074627     Poe
11            cajun    positive     12            0.059701     Poe
3   africanamerican    positive     11            0.054726     Poe
47          kurdish    positive      9            0.044776     Poe
39         japanese    positive      8            0.039801     Poe
15           celtic    positive      7            0.034826     Poe
26            greek    positive      7            0.034826     Poe
58          russian    positive      7            0.034826     Poe
31         hispanic    positive      6            0.029851     Poe
66             sikh    positive      6            0.029851     Poe
56          punjabi    positive      5            0.024876     Poe
7              arab    positive      3            0.014925     Poe
19            cuban    positive      3            0.014925     Poe
37          italian    positive      3            0.014925     Poe
52   nativeamerican    positive      3            0.014925     Poe
63           indian    positive     79            0.393035  Austen
3           african    positive     15            0.074627  Austen
25          english    positive     15            0.074627  Austen
33    englishindian    positive     14            0.069652  Austen
74          kurdish    positive     13            0.064677  Austen
22             cree    positive      9            0.044776  Austen
19        caucasian    positive      7            0.034826  Austen
65            irish    positive      6            0.029851  Austen
49   englishmestiza    positive      3            0.014925  Austen
81   nativeamerican    positive      3            0.014925  Austen
93         scottish    positive      3            0.014925  Austen

In [18]:
result = chi_square_test_authors(df_concat, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(209.7813),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 24}

In [19]:
result = compute_standardized_residuals_authors(df_concat, index_col='ethnicity')
result

author             Austen       Poe
ethnicity                          
african          2.811862 -2.762671
africanamerican -2.324423  2.283759
cajun           -2.427780  2.385309
cree             2.178059 -2.139956
english          2.811862 -2.762671
englishindian    2.716517 -2.668994
french-canadian -2.803359  2.754317
indian           4.263792 -4.189201
persian         -2.714341  2.666856

In [20]:
authors = df_concat["author"].unique()
result = compute_divergences_authors(df_concat, authors, index_col='ethnicity')
result

{'JSD': {'JSD(Poe || Austen)': np.float64(0.7673080624812966)},
 'KL': {'KL(Poe || Austen)': np.float64(12.600899879532538),
  'KL(Austen || Poe)': np.float64(7.803355110016076)}}

### Negative - Ethnicity

In [21]:
poe_neg = poe[poe['connotation'] == 'negative']
austen_neg = austen[austen['connotation'] == 'negative']

In [22]:
df_list = [poe_neg, austen_neg]
df_concat = pd.concat(df_list)
df_concat

ethnicity connotation  count  relative_frequency  author
34            irish    negative     88            0.437811     Poe
12        caucasian    negative     31            0.154229     Poe
2   africanamerican    negative     21            0.104478     Poe
10            cajun    negative      9            0.044776     Poe
46          kurdish    negative      9            0.044776     Poe
24  french-canadian    negative      4            0.019900     Poe
1           african    negative      3            0.014925     Poe
6              arab    negative      3            0.014925     Poe
14           celtic    negative      3            0.014925     Poe
28            gypsy    negative      3            0.014925     Poe
69          spanish    negative      3            0.014925     Poe
62           indian    negative     74            0.185000  Austen
2           african    negative     62            0.155000  Austen
24          english    negative     56            0.140000  Austen
32    englishindian    negative     27            0.067500  Austen
18        caucasian    negative     24            0.060000  Austen
64            irish    negative     21            0.052500  Austen
73          kurdish    negative     19            0.047500  Austen
92         scottish    negative     10            0.025000  Austen
0            afghan    negative      8            0.020000  Austen
4   africanamerican    negative      8            0.020000  Austen
9             black    negative      6            0.015000  Austen
85         romanian    negative      6            0.015000  Austen
94   scottishgaelic    negative      6            0.015000  Austen
15    britishindian    negative      4            0.010000  Austen
71           korean    negative      4            0.010000  Austen

In [23]:
result = chi_square_test_authors(df_concat, index_col='ethnicity')
result

{'Chi-Square Statistic': np.float64(311.9793),
 'p-value': np.float64(0.0),
 'Degrees of Freedom': 20}

In [24]:
result = compute_standardized_residuals_authors(df_concat, index_col='ethnicity')
result

author             Austen       Poe
ethnicity                          
african          2.985641 -4.107460
africanamerican -2.519432  3.466079
cajun           -2.426659  3.338447
english          3.198231 -4.399929
englishindian    2.220740 -3.055157
indian           3.676476 -5.057868
irish           -5.958347  8.197125

In [25]:
authors = df_concat["author"].unique()
result = compute_divergences_authors(df_concat, authors, index_col='ethnicity')
result

{'JSD': {'JSD(Poe || Austen)': np.float64(0.7804045552974412)},
 'KL': {'KL(Poe || Austen)': np.float64(4.068909359970863),
  'KL(Austen || Poe)': np.float64(12.68057068112229)}}